app.py

```diff
@@ -742,8 +742,20 @@ def newlabel():
 	orig_senttok, _ = worker.postokenize(sent)
 	if app.config['CGELVALIDATE'] is None:
 		treestr = request.args.get('tree')
-	else: 
-		treestr = "(ROOT " + cgel.parse(request.args.get('tree'))[0].ptb() + ")"
+	else:
+		cgel_tree = cgel.parse(request.args.get('tree'))[0]
+		tags = {}
+		keys = range(1,len(orig_senttok)+1)
+		cgel_tree_terminals = [node for node in cgel_tree.tokens.values() if node.text or node.constituent == 'GAP']
+		for i in keys:
+			tags[i] = {"note": None, "xpos" : None, "_lemma" : None}
+			if cgel_tree_terminals[i-1].note:
+				tags[i].update({"note": cgel_tree_terminals[i-1].note})
+			if cgel_tree_terminals[i-1].xpos:
+				tags[i].update({"xpos": cgel_tree_terminals[i-1].xpos})
+			if cgel_tree_terminals[i-1]._lemma:
+				tags[i].update({"_lemma": cgel_tree_terminals[i-1]._lemma})
```

Comment:
I now define an array of terminal nodes `cgel_tree_terminals = cgel_tree.terminals(gaps=True)`. Later on, I attempt to modify the `terminals` property of a cgel tree object `treestr` by running: `treestr.terminals = cgel_tree_terminals`. This does indeed update the `treestr.terminals` but does not change the output of `str(treestr)`, i.e., `str(treestr)` doesn't include the tags of nodes in `cgel_tree_terminals`.

Reply:
Expected behavior (from the debug console):
`> treestr.terminals[0].xpos
'test'
> str(treestr)
'(Clause\n    :Subj (NP\n        :Det (DP\n            :Head (D :t "The" :xpos "test"))\n        :Head (Nom\n            :Head (Adv :t "girl")))\n    :Head (VP\n        :Head (V :t "ate")\n        :Obj (NP\n            :Head (Nom\n                :Head (N :t "pizza")))\n        :Comp (PP\n            :Head (P :t "with")\n            :Obj (NP\n                :Head (Nom\n                    :Head (N :t "tomatoes"))))))'`

actual behavior:
`> treestr.terminals[0].xpos
'test'
> str(treestr)
'(Clause\n    :Subj (NP\n        :Det (DP\n            :Head (D :t "The"))\n        :Head (Nom\n            :Head (Adv :t "girl")))\n    :Head (VP\n        :Head (V :t "ate")\n        :Obj (NP\n            :Head (Nom\n                :Head (N :t "pizza")))\n        :Comp (PP\n            :Head (P :t "with")\n            :Obj (NP\n                :Head (Nom\n                    :Head (N :t "tomatoes"))))))'`

Reply:
That's because `terminals` is a method not a property. You can't assign to it.

I think adding 2 methods to cgel.py, `update_terminals()` and recursive helper `_update_terminals_rec()`, would be the simplest solution. These would be like `terminals()` and `_terminals_rec()` but instead of adding terminals to a list, `_update_terminals_rec()` would set `self.tokens[cur]` to a new value. I can work on this later tonight if that would help.

Reply:
Oh, right! Yeah then I think cgel tree objects do need a method for updating terminals

Reply:
Haven't tested it but I think this will work: https://github.com/nert-nlp/cgel/commit/44ee8567d781ceba534b6702a68b3b5b847224e8

Reply:
TBC the idea is to regenerate the CGEL tree data structure from the visual tree export, then restore the original terminal nodes including any extraneous attributes like notes that were not part of the visual tree, but substitute new category and function for those terminal nodes?

Reply:
(should I add options in `update_terminals()` to keep the category and function while otherwise updating the node?)

Reply:
> TBC the idea is to regenerate the CGEL tree data structure from the visual tree export, then restore the original terminal nodes including any extraneous attributes like notes that were not part of the visual tree, but substitute new category and function for those terminal nodes?

Yes, exactly!

> (should I add options in `update_terminals()` to keep the category and function while otherwise updating the node?)

I think this would make things very convenient on my end!

Reply:
OK how about this: https://github.com/nert-nlp/cgel/commit/f51386a3693869fc2e35964d040faf34a43f73a8

Reply:
Just tried this and getting an error:
```
File "/Users/bw/Documents/GitHub/activedop/cgel/cgel.py", line 509, in update_terminals
    oldterm = self.tokens[i]
              ^^^^^^^^^^^^^^
KeyError: <cgel.Node object at 0x12dda8ce0>
```
I think this is because `i` needs to be an integer (a key within `self.tokens`?)

Reply:
Fixed—it was wrong recursive call. (I'm surprised the type-checker didn't catch it. It caught it after after declaring an explicit type for `result`.)

Reply:
Awesome, working great now as far as I can see!